AssertReleaseable failures shall show up details why not releaseable

## Situation
We have the problem that it is not clear why the pds-tools release 1.1.0 cannot be done because the `assertReleasable` gradle task fails with a the error message (see Details), but none of the 
versions is marked as `dirty` - so the assert shall not throw a failure!

## Wanted
It shall be clear which version element/s is/are responsible and why.
This details shall be shown in case of a failure at assertion time.


### Details
#### Action console output
```
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
> Task :buildSrc:compileJava NO-SOURCE
> Task :buildSrc:compileGroovy UP-TO-DATE
> Task :buildSrc:processResources NO-SOURCE
> Task :buildSrc:classes UP-TO-DATE
> Task :buildSrc:jar UP-TO-DATE

> Configure project :
BUILD versioning
- Client  :0.0.0 [0.0.0-369be7f-20231025160905]
- Server  :0.0.0 [0.0.0-369be7f-20231025160905]
- PDS     :0.0.0 [0.0.0-369be7f-20231025160905]
- PDS-Tools :1.1.0 [1.1.0-20231025160905]
- Libraries :0.0.0 [0.0.0-369be7f-20231025160905]
- Checkmarx Wrapper :0.0.0 [0.0.0-369be7f-20231025160905]
- OWASP-ZAP Wrapper :0.0.0 [0.0.0-369be7f-20231025160905]

Time elapsed to build versions: 0.273 seconds

> Configure project :sechub-cli
publish sechub-cli will be ignored at github actions at the moment!


FAILURE: Build failed with an exception.

* Where:
Script '/home/runner/work/sechub/sechub/gradle/build-versioning.gradle' line: 213

* What went wrong:
Execution failed for task ':assertReleaseable'.
> At least one version contains uncommitted changes and is marked for release! See git-status.txt artifact.
```
#### Attached file git-status.txt
```
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

--- a/.github/workflows/release-pds-tools.yml
+++ b/.github/workflows/release-pds-tools.yml
@@ -99,6 +99,7 @@ jobs:
           git restore sechub-api-java/src/main/resources/reduced-openapi3.json
           mkdir -p build/reports
           git status > build/reports/git-status.txt
+          git describe --tags >> build/reports/git-status.txt
 
       - name: Archive GIT status
         if: always()

--- a/.github/workflows/release-pds-tools.yml
+++ b/.github/workflows/release-pds-tools.yml
@@ -99,7 +99,8 @@ jobs:
           git restore sechub-api-java/src/main/resources/reduced-openapi3.json
           mkdir -p build/reports
           git status > build/reports/git-status.txt
-          git describe --tags >> build/reports/git-status.txt
+          echo "--- git tags:" >> build/reports/git-status.txt
+          git tag --points-at HEAD >> build/reports/git-status.txt
 
       - name: Archive GIT status
         if: always()

--- a/buildSrc/src/main/groovy/VersionData.groovy
+++ b/buildSrc/src/main/groovy/VersionData.groovy
@@ -9,6 +9,7 @@ class VersionData{
     private static final String ID_WEBSITE = "website"
     private static final String ID_WRAPPER_CHECKMARX = "checkmarx wrapper"
     private static final String ID_WRAPPER_OWASPZAP = "owasp-zap wrapper"
+    private static final StringBuilder debugInfo = new StringBuilder();
 
     private static Map<String,VersionInfo> map = new HashMap<>();
 
@@ -56,7 +57,7 @@ class VersionData{
         if (info==null){
             throw new IllegalArgumentException("unsupported version type:"+versionType);
         }
-        inspectReleaseVersion(fullVersion);
+        inspectReleaseVersion(versionType, fullVersion);
         info.shortVersion = simplifiedVersion(fullVersion);
         info.fullVersion= fullVersion
 
@@ -98,18 +99,30 @@ class VersionData{
     public static String getWebsiteVersion(){
         return map.get(ID_WEBSITE).getShortVersion()
     }
+    
+    public static String getDebugInfo(){
+        
+        return "Debug info:\ncontainingAtLeastOneDirtyReleaseVersion=$containingAtLeastOneDirtyReleaseVersion\ncontainingAtLeastOneRealReleaseVersion=$containingAtLeastOneRealReleaseVersion\n\n$debugInfo";                
+    }
+
 
     /**
      * Inspect version - if not starting with 0.0.0 this means it's a release, so
      *                   a "dirty" may not be contained inside long version name
      */
-    private static void inspectReleaseVersion(String longVersionName){
+    private static void inspectReleaseVersion(String versionType, String longVersionName){
+        debugInfo.append("\ninspect $versionType release version: long version=$longVersionName\n")
+        debugInfo.append("- at least one release found : $containingAtLeastOneRealReleaseVersion, one release dirty: $containingAtLeastOneDirtyReleaseVersion\n")
+        
         if (longVersionName.startsWith("0.0.0")){
             /* not a correct release version so ignore */
             return
         }
         containingAtLeastOneDirtyReleaseVersion=containingAtLeastOneDirtyReleaseVersion || longVersionName.contains("dirty")
         containingAtLeastOneRealReleaseVersion=true
+        
+        debugInfo.append("- updated data")
+        debugInfo.append("- at least one release found : $containingAtLeastOneRealReleaseVersion, one release dirty: $containingAtLeastOneDirtyReleaseVersion\n")
     }
 
     /**

--- a/gradle/build-versioning.gradle
+++ b/gradle/build-versioning.gradle
@@ -210,10 +210,10 @@ task assertReleaseable {
     doLast {
         /* we only assert at execution phase - means when this task is really executed and not only configured...*/
         if (VersionData.isContainingAtLeastOneDirtyReleaseVersion()){
-            throw new IllegalStateException("At least one version contains uncommitted changes and is marked for release! See git-status.txt artifact.")
+            throw new IllegalStateException("At least one version contains uncommitted changes and is marked for release! See git-status.txt artifact.\n"+VersionData.getDebugInfo())
         }
         if (!VersionData.isContainingAtLeastOneRealReleaseVersion()){
-            throw new IllegalStateException("There is not at least one version defined which can be released! 0.0.0 is not an acceptable release version!")
+            throw new IllegalStateException("There is not at least one version defined which can be released! 0.0.0 is not an acceptable release version!\n"+VersionData.getDebugInfo())
         }
     }
 

